httpx/middleware/auth.py

```diff
@@ -0,0 +1,219 @@
+import hashlib
+import os
+import time
+import typing
+from base64 import b64encode
+from urllib.request import parse_http_list
+
+from ..exceptions import ProtocolError
+from ..models import AsyncRequest, AsyncResponse, StatusCode
+from ..utils import to_bytes, to_str, unquote
+from .base import BaseMiddleware
+
+
+class BasicAuth(BaseMiddleware):
+    def __init__(
+        self, username: typing.Union[str, bytes], password: typing.Union[str, bytes]
+    ):
+        if isinstance(username, str):
+            username = to_bytes(username)
+
+        if isinstance(password, str):
+            password = to_bytes(password)
```

Comment:
Sorry to nitpick but I think the instance check is already made by to_bytes(), isn’t it?

If mypy complains then perhaps we can consider converting to bytes directly when building the `userpass` below?